django/core/servers/basehttp.py

```diff
@@ -144,7 +144,7 @@ def cleanup_headers(self):
             self.request_handler.close_connection = True
 
     def close(self):
-        self.get_stdin()._read_limited()
```

Comment:
Not your fault (ie this issue already exists), but this code does *not* exhaust the stream; we'd need a while loop here till `read()` returns "" or add an exhaust method like werkzeug

Reply:
Probably more important, this means that https://github.com/django/django/blob/cb791a2540c289390b68a3ea9c6a79476890bab2/django/http/request.py#L338 is not guaranteed to read the whole body data. I think we should copy more from werkzeug (they are failing if not everything is readable)

Reply:
Yes. I sort of noticed this. Will implement something here like `werkzeug` does.

Reply:
So looking at this a bit more. In `werkzeug` they [check if beyond the limit](https://github.com/pallets/werkzeug/blob/dbf78f67/src/werkzeug/wsgi.py#L915-L916) and call [`on_exhausted()`](https://github.com/pallets/werkzeug/blob/dbf78f67/src/werkzeug/wsgi.py#L876-L883) which calls `self._read(0)` to ensure an `b""` is returned. My implementation handles this by returning `b""` directly which is based on the original behaviour of what we had in `_read_limited()`.

They also have the `exhaust()` method which is called in [`exhaust_stream()`](https://github.com/pallets/werkzeug/blob/a3b4572a34269efaca4d91fa4cd07dd7f6f94b6d/src/werkzeug/formparser.py#L135-L153), but that also has handling for streams that don't have an `exhaust()`. I guess this special implementation is needed as we need to bypass the limit and read everything that remains beyond the limit. Will add a separate commit for this.

Reply:
No. Wait. Am confusing myself. They only ever read up to the limit in `werkzeug` as `exhaust()` calls `self.read()` and not `self._read()` (the `read()` of the wrapped stream object).

Also, while a while loop may make sense, I think that we're probably fine without it. We're calling `read()` with no `size` argument in these cases. That means we'll end up with `size = limit - _pos` which means we'll read everything  we need to in both of the cases highlighted above. It's not ideal as we should probably have a subsequent call that returns `b""`, but it's been working up to now... I think we can punt any review of this to a separate ticket.

Reply:
Not 100% sure of the conclusion here.

Reply:
I think that we can just leave this for now and follow up in a separate ticket if we think there is something to do here - I'm not sure. We're no worse off than we were.